src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                if (!CanTryConvertToLocalFunction() ||
+                    !TryCreateStackFromQueryExpression(out QueryExpressionProcessingInfo queryExpressionProcessingInfo))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if ((TryConvertInternal(queryExpressionProcessingInfo, out documentUpdateInfo) ||
+                    TryReplaceWithLocalFunction(out documentUpdateInfo)) && // second attempt: at least to a local function
+                    !_semanticModel.GetDiagnostics(_source.Span, _cancellationToken).Any())
+                {
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private StatementSyntax ProcessClause(CSharpSyntaxNode node, StatementSyntax statement, out StatementSyntax extraStatementToAddAbove)
+            {
+                extraStatementToAddAbove = default;
+                switch (node.Kind())
+                {
+                    case SyntaxKind.WhereClause:
+                        return SyntaxFactory.Block(SyntaxFactory.IfStatement(((WhereClauseSyntax)node).Condition.WithAdditionalAnnotations(Simplifier.Annotation).WithoutTrivia(), statement));
+                    case SyntaxKind.FromClause:
+                        var fromClause = (FromClauseSyntax)node;
+                        return SyntaxFactory.ForEachStatement(
+                                    fromClause.Type ?? VarNameIdentifier,
+                                    fromClause.Identifier,
+                                    fromClause.Expression.WithoutTrivia(), WrapWithBlock(statement));
+                    case SyntaxKind.LetClause:
+                        var letClause = (LetClauseSyntax)node;
+                        return AddToBlockTop(CreateLocalDeclarationStatement(letClause.Identifier, letClause.Expression), statement);
+                    case SyntaxKind.JoinClause:
+                        var joinClause = (JoinClauseSyntax)node;
+                        if (joinClause.Into != null)
+                        {
+                            // This must be caught on the validation step. Therefore, here is an exception.
+                            throw new ArgumentException("GroupJoin is not supported");
+                        }
+                        else
+                        {
+                            ExpressionSyntax expression;
+                            if (IsLocalOrParameterSymbol(_semanticModel.GetOperation(joinClause.InExpression, _cancellationToken)))
```

Comment:
>IsLocalOrParameterSymbol(_semanticModel.GetOperation(joinClause.InExpression, _cancellationToken)) [](start = 32, length = 98)

This feels more complicated then it could be. I would just dig through parenthesis and call SemanticModel.GetSymbolInfor and check the symbol. BTW, should probably allow range variable symbol here as well.